go/batch-submitter/drivers/proposer/driver.go

```diff
@@ -0,0 +1,168 @@
+	CTCAddr     common.Address
+	ChainID     *big.Int
+	PrivKey     *ecdsa.PrivateKey
+}
+
+type Driver struct {
+	cfg         Config
+	sccContract *scc.StateCommitmentChain
+	ctcContract *ctc.CanonicalTransactionChain
+	walletAddr  common.Address
+}
+
+func NewDriver(cfg Config) (*Driver, error) {
+	sccContract, err := scc.NewStateCommitmentChain(
+		cfg.SCCAddr, cfg.L1Client,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	ctcContract, err := ctc.NewCanonicalTransactionChain(
+		cfg.CTCAddr, cfg.L1Client,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	walletAddr := crypto.PubkeyToAddress(cfg.PrivKey.PublicKey)
+
+	return &Driver{
+		cfg:         cfg,
+		sccContract: sccContract,
+		ctcContract: ctcContract,
+		walletAddr:  walletAddr,
+	}, nil
+}
+
+// Name is an identifier used to prefix logs for a particular service.
+func (d *Driver) Name() string {
+	return d.cfg.Name
+}
+
+// WalletAddr is the wallet address used to pay for batch transaction fees.
+func (d *Driver) WalletAddr() common.Address {
+	return d.walletAddr
+}
+
+// GetBatchBlockRange returns the start and end L2 block heights that need to be
+// processed. Note that the end value is *exclusive*, therefore if the returned
+// values are identical nothing needs to be processed.
+func (d *Driver) GetBatchBlockRange(
+	ctx context.Context) (*big.Int, *big.Int, error) {
+
+	blockOffset := new(big.Int).SetUint64(d.cfg.BlockOffset)
+	maxBatchSize := new(big.Int).SetUint64(1)
+
+	start, err := d.sccContract.GetTotalElements(&bind.CallOpts{
+		Pending: false,
+		Context: ctx,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	start.Add(start, blockOffset)
+
+	totalElements, err := d.ctcContract.GetTotalElements(&bind.CallOpts{
+		Pending: false,
+		Context: ctx,
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+	totalElements.Add(totalElements, blockOffset)
+
+	// Take min(start + blockOffset + maxBatchSize, totalElements).
+	end := new(big.Int).Add(start, maxBatchSize)
+	if totalElements.Cmp(end) < 0 {
+		end.Set(totalElements)
+	}
+
+	if start.Cmp(end) > 0 {
+		return nil, nil, fmt.Errorf("invalid range, "+
+			"end(%v) < start(%v)", end, start)
+	}
+
+	return start, end, nil
+}
+
+// SubmitBatchTx transforms the L2 blocks between start and end into a batch
+// transaction using the given nonce and gasPrice. The final transaction is
+// published and returned to the call.
+func (d *Driver) SubmitBatchTx(
+	ctx context.Context,
+	start, end, nonce, gasPrice *big.Int) (*types.Transaction, error) {
+
+	var blocks []*l2types.Block
+	for i := new(big.Int).Set(start); i.Cmp(end) < 0; i.Add(i, bigOne) {
+		block, err := d.cfg.L2Client.BlockByNumber(ctx, i)
+		if err != nil {
+			return nil, err
+		}
+
+		blocks = append(blocks, block)
+
+		// TODO(conner): remove when moving to multiple blocks
+		break //nolint
+	}
+
+	var stateRoots = make([][32]byte, 0, len(blocks))
+	for _, block := range blocks {
+		stateRoots = append(stateRoots, block.Root())
+	}
+
+	signerFn := func(
```

Comment:
Unsure if you know that this exists: https://github.com/ethereum/go-ethereum/blob/cc87cbd70af18241f79908eec10bb3c51a9fbda4/accounts/abi/bind/auth.go#L145

This could build the `TransactOpts` to which you could modify the nonce/ctx after the fact, doing it the way you are doing it is fine too, it would just standardize a bit between services

Reply:
great suggestion, fixed!